packages/core/versioning/core-versioned-http-server-internal/src/internal_versioned_route.ts

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { schema } from '@kbn/config-schema';
+import type {
+  RequestHandler,
+  IRouter,
+  RequestHandlerContextBase,
+  KibanaRequest,
+  KibanaResponseFactory,
+} from '@kbn/core-http-server';
+import type {
+  Version,
+  AddVersionOpts,
+  VersionedRoute,
+  VersionedRouteConfig,
+} from '@kbn/core-versioned-http-server';
+import type { CoreKibanaRequest } from '@kbn/core-http-router-server-internal';
+import type { Method } from './types';
+
+import { validate } from './validate';
+
+type Options = AddVersionOpts<unknown, unknown, unknown, unknown>;
+
+export const VERSION_HEADER = 'TBD';
+
+// This validation is a pass-through so that we can apply our version-specific validation later
+const passThroughValidation = { body: schema.any(), params: schema.any(), query: schema.any() };
+
+export class InternalVersionedRoute implements VersionedRoute {
+  private readonly handlers = new Map<
+    Version,
+    {
+      fn: RequestHandler;
+      options: Options;
+    }
+  >();
+
+  public static from({
+    router,
+    method,
+    path,
+    options,
+  }: {
+    router: IRouter;
+    method: Method;
+    path: string;
+    options: VersionedRouteConfig<Method>;
+  }) {
+    return new InternalVersionedRoute(router, method, path, options);
+  }
+
+  private constructor(
+    private readonly router: IRouter,
+    public readonly method: Method,
+    public readonly path: string,
+    public readonly options: VersionedRouteConfig<Method>,
+    // TODO: Make "true" dev-only
+    private readonly validateResponses: boolean = true
+  ) {}
+
+  /** This is where we must implement the versioned spec once it is available */
+  private requestHandler = async (
+    ctx: RequestHandlerContextBase,
+    req: KibanaRequest,
+    res: KibanaResponseFactory
+  ) => {
+    const version = req.headers[VERSION_HEADER] as undefined | Version;
+    if (!version) {
+      return res.custom({
+        statusCode: 406,
+        body: `Version expected at [${this.method}] [${this.path}]. Please specify a version in the ${VERSION_HEADER} header.`,
+      });
+    }
+
+    const handler = this.handlers.get(version);
+    if (!handler) {
+      return res.custom({
+        statusCode: 406,
+        body: `No ${version} available for ${this.method} ${this.path}. Available versions are: ${[
+          this.handlers.keys(),
+        ].join(',')}`,
+      });
+    }
+
+    const coreKibanaRequest = req as CoreKibanaRequest;
+    if (handler.options.validate && handler.options.validate.request) {
```

Comment:
NIT: `if (handler.options.validate?.request)`

Reply:
That won't work because at this point `validate` could be `false`.